### PR TITLE
Add e2e testing with uptest

### PR DESCRIPTION
**Pull request by turkenh**
### Description of your changes

End to end testing with uptest.

I have:

- [x] Read and followed Upbound's [contribution process](https://git.io/fj2m9).
- [ ] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR, as appropriate.

### How has this code been tested

```shell
export UPTEST_GCP_CREDS=$(cat ~/Desktop/my-gcp-sa.json)
export UPTEST_GCP_PROJECT=<my-gcp-project>

make e2e
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -48,4 +48,37 @@ submodules:
...
+
+controlplane.down: $(UP) $(KUBECTL) $(KIND)
+	@$(INFO) deleting controlplane
+	@$(KIND) get kubeconfig --name $(KIND_CLUSTER_NAME) >/dev/null 2>&1 && $(KIND) delete cluster --name=$(KIND_CLUSTER_NAME)
+	@$(OK) deleting controlplane
+
+uptest-local: $(UP) $(KUBECTL) $(KUTTL)
+	@$(INFO) running automated tests
+	@$(KUBECTL) apply -R -f package/cluster
+	@KUBECTL=$(KUBECTL) KUTTL=$(KUTTL) uptest --claim-or-composite --example-list=examples/cluster-claim.yaml --default-timeout=2400 || $(FAIL)
```
Comment (turkenh) [inline Makefile:81]: TODO(turkenh): This line needs to be replaced with available `uptest` binary.

File: Makefile [hunk 2]
```diff
@@ -48,4 +48,37 @@ submodules:
...
+PROVIDER_GCP_VERSION ?= v0.15.0
+PROVIDER_HELM_VERSION ?= v0.12.0
+
+controlplane.up: $(UP) $(KUBECTL) $(KIND)
+	@$(INFO) setting up controlplane
+	@$(KIND) get kubeconfig --name $(KIND_CLUSTER_NAME) >/dev/null 2>&1 || $(KIND) create cluster --name=$(KIND_CLUSTER_NAME)
+	@$(KUBECTL) -n upbound-system get cm universal-crossplane-config >/dev/null 2>&1 || $(UP) uxp install
+	@$(KUBECTL) -n upbound-system wait deploy crossplane --for condition=Available --timeout=120s
+	@$(KUBECTL) -n upbound-system create secret docker-registry package-pull-secret --docker-server=xpkg.upbound.io --docker-username=${UPBOUND_DOCKER_USERNAME} --docker-password=${UPBOUND_DOCKER_PASSWORD} -o yaml --dry-run=client | $(KUBECTL) apply -f -
+	@$(KUBECTL) get provider.pkg upbound-provider-gcp > /dev/null 2>&1 || $(UP) ctp provider install upbound/provider-gcp:$(PROVIDER_GCP_VERSION) --package-pull-secrets=package-pull-secret
```
Comment (ytsarev) [inline Makefile:67]: all logic around `package-pull-secret` can be removed as we are not gating official providers anymore

**Review by ytsarev — CHANGES_REQUESTED** [review]

File: Makefile (hunk 1 above)
Comment (ytsarev) [reply Makefile:81]: Should we add `uptest` to https://github.com/upbound/build?

Comment (ytsarev) [reply Makefile:81]: @turkenh where can I get the correct `uptest` binary version for local `uptest-local` testing?

Comment (turkenh) [reply Makefile]: It is now part of build submodule, you should be able to checkout the PR and run `make e2e` after setting up your creds with UPTEST_GCP_CREDS env var.

File: Makefile [hunk 3]
```diff
@@ -48,4 +48,36 @@ submodules:
...
+	@$(KUBECTL) wait provider.pkg upbound-provider-gcp --for condition=Healthy --timeout=120s
+	@$(KUBECTL) wait provider.pkg crossplane-contrib-provider-helm --for condition=Healthy --timeout=120s
+	@$(OK) setting up controlplane
+
+controlplane.down: $(UP) $(KUBECTL) $(KIND)
+	@$(INFO) deleting controlplane
+	@$(KIND) get kubeconfig --name $(KIND_CLUSTER_NAME) >/dev/null 2>&1 && $(KIND) delete cluster --name=$(KIND_CLUSTER_NAME)
+	@$(OK) deleting controlplane
+
+uptest-local: $(UPTEST) $(KUBECTL) $(KUTTL)
```
Comment (muvaf) [inline Makefile:77]:
```suggestion
uptest: $(UPTEST) $(KUBECTL) $(KUTTL)
```

It's not doing anything specific to local anymore, right?

File: .github/workflows/e2e.yaml [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+      - name: Detect No-op Changes
+        id: noop
+        uses: fkirc/skip-duplicate-actions@v2.0.0
+        with:
+          github_token: ${{ secrets.GITHUB_TOKEN }}
+          paths_ignore: '["**.md", "**.png", "**.jpg"]'
+          do_not_skip: '["workflow_dispatch", "schedule", "push"]'
+
+  e2e:
+    runs-on: ubuntu-20.04
```
Comment (muvaf) [inline .github/workflows/e2e.yaml:29]:
```suggestion
    runs-on: ubuntu-22.04
```

File: Makefile [hunk 5]
```diff
@@ -48,4 +48,36 @@ submodules:
...
+KIND_CLUSTER_NAME ?= uptest
+PROVIDER_GCP_VERSION ?= v0.15.0
+PROVIDER_HELM_VERSION ?= v0.12.0
+
+controlplane.up: $(UP) $(KUBECTL) $(KIND)
+	@$(INFO) setting up controlplane
+	@$(KIND) get kubeconfig --name $(KIND_CLUSTER_NAME) >/dev/null 2>&1 || $(KIND) create cluster --name=$(KIND_CLUSTER_NAME)
+	@$(KUBECTL) -n upbound-system get cm universal-crossplane-config >/dev/null 2>&1 || $(UP) uxp install
+	@$(KUBECTL) -n upbound-system wait deploy crossplane --for condition=Available --timeout=120s
+	@$(KUBECTL) get provider.pkg upbound-provider-gcp > /dev/null 2>&1 || $(UP) ctp provider install upbound/provider-gcp:$(PROVIDER_GCP_VERSION)
```
[the remark this answers is not in the document]
Comment (turkenh) [reply Makefile:66]: Yes, actually have a draft PR in build submodule: https://github.com/upbound/build/pull/208/files

I remember, I tried to make it work for provider packages as well and gave up after seeing things getting too complicated. But I will give it another try only with configuration package support (for now).

File: .github/workflows/e2e.yaml [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        env:
+          GITHUB_TOKEN: '${{ secrets.GITHUB_TOKEN }}'
+
+      - name: End to end
+        run: make e2e
+        env:
+          UPTEST_GCP_PROJECT: ${{ secrets.UPTEST_GCP_PROJECT }}
+          UPTEST_GCP_CREDS: ${{ secrets.UPTEST_GCP_CREDS }}
+          UPBOUND_DOCKER_USERNAME: ${{ secrets.UPBOUND_DOCKER_USERNAME }}
+          UPBOUND_DOCKER_PASSWORD: ${{ secrets.UPBOUND_DOCKER_PASSWORD }}
```
[the remark this answers is not in the document]
Comment (turkenh) [reply .github/workflows/e2e.yaml:57]: Nope, legacy.

File: Makefile [hunk 7]
```diff
@@ -48,4 +48,36 @@ submodules:
...
+KIND_VERSION = v0.16.0
+KIND_CLUSTER_NAME ?= uptest
+PROVIDER_GCP_VERSION ?= v0.15.0
+PROVIDER_HELM_VERSION ?= v0.12.0
+
+controlplane.up: $(UP) $(KUBECTL) $(KIND)
+	@$(INFO) setting up controlplane
+	@$(KIND) get kubeconfig --name $(KIND_CLUSTER_NAME) >/dev/null 2>&1 || $(KIND) create cluster --name=$(KIND_CLUSTER_NAME)
+	@$(KUBECTL) -n upbound-system get cm universal-crossplane-config >/dev/null 2>&1 || $(UP) uxp install
+	@$(KUBECTL) -n upbound-system wait deploy crossplane --for condition=Available --timeout=120s
```
[the remark this answers is not in the document]
Comment (turkenh) [reply Makefile]: Moved: https://github.com/upbound/build/pull/208

File: Makefile (hunk 5 above)
[the remark this answers is not in the document]
Comment (turkenh) [reply Makefile]: Ok, looks like we are good to go now, especially with the removal of package pull secret requirement things are more straightforward. 
Also to indicate only xpkg deployment is supported, I have named the target as `local.xpkg.deploy-cfg.<package>` which made me feel better: https://github.com/upbound/build/pull/208

For the changes on this side, please see the last commit.

**Review by muvaf — APPROVED** [review]
Thanks @turkenh ! Looking great!

File: Makefile [hunk 8]
```diff
@@ -26,6 +26,7 @@ XPKGS = $(PROJECT_NAME)
 
 CROSSPLANE_NAMESPACE = upbound-system
 -include build/makelib/local.xpkg.mk
+-include build/makelib/controlplane.mk
```
Comment (jbw976) [inline Makefile:29]: where does this `controlplane.mk` file live? i didn't see it in https://github.com/upbound/build/tree/master/makelib

File: test/setup.sh [hunk 9]
```diff
@@ -0,0 +1,29 @@
+#!/usr/bin/env bash
+set -aeuo pipefail
+
+UPTEST_GCP_PROJECT=${UPTEST_GCP_PROJECT:-crossplane-playground}
+
+echo "Running setup.sh"
+echo "Waiting until configuration package is healthy/installed..."
+${KUBECTL} wait configuration.pkg platform-ref-gcp --for=condition=Healthy --timeout 5m
```
[the remark this answers is not in the document]
Comment (turkenh) [reply test/setup.sh:8]: it is a dependency to `uptest` target with `local.xpkg.deploy-cfg.$(PROJECT_NAME)`

[the remark this answers is not in the document]
Comment (turkenh) [reply test/setup.sh:8]: It is also part of the same build pr: https://github.com/upbound/build/pull/208

**Review by ytsarev — APPROVED** [review]
LGTM

**Pull request merged**